apps/src/sites/code.org/pages/public/learn/local.js

```diff
@@ -111,37 +124,219 @@ function displayQueryError() {
 }
 
 function loadMap(locations) {
-  var coordinates = gmap_loc.split(',');
+  var coordinates = map_loc.split(',');
   var lat = coordinates[0];
   var lng = coordinates[1];
 
   // Reset the map.
-  $('#gmap').html('');
-  gmap = new Maplace();
-
-  var map_options = {
-    map_options: {
-      set_center: [lat, lng],
-      zoom: 12
-    },
-    controls_type: 'list',
-    controls_on_map: false
-  };
+  $('#map').html('');
+  if (window.location.search.includes('mapbox')) {
+    const featureList = locations.map((location, index) => ({
+      type: 'Feature',
+      properties: {
+        description: location.html,
+        title: location.title,
+        index
+      },
+      geometry: {
+        type: 'Point',
+        coordinates: [parseFloat(location.lon), parseFloat(location.lat)]
+      }
+    }));
+    map = new mapboxgl.Map({
+      container: 'map',
+      style: 'mapbox://styles/mapbox/streets-v11',
+      zoom: 10,
+      minZoom: 1,
+      center: [lng, lat]
+    });
+    map.on('load', function() {
+      map.loadImage(
+        'https://docs.mapbox.com/mapbox-gl-js/assets/custom_marker.png',
+        function(error, image) {
+          if (error) {
+            logToCloud.addPageAction(
+              logToCloud.PageAction.MapboxMarkerLoadError,
+              {
+                error
+              }
+            );
+            throw error;
+          }
+          map.addImage('custom-marker', image);
+          map.addSource('places', {
+            type: 'geojson',
+            data: {
+              type: 'FeatureCollection',
+              features: featureList
+            }
+          });
+          // Add a layer showing the places.
+          map.addLayer({
+            id: 'places',
+            type: 'symbol',
+            source: 'places',
+            layout: {
+              'icon-image': 'custom-marker',
+              'icon-allow-overlap': true
+            }
+          });
 
-  if (locations.length > 0) {
-    map_options.force_generate_controls = true;
-    map_options.locations = locations;
-    map_options.afterOpenInfowindow = function(index, location, marker) {
-      setDetailsTrigger(index, location, marker);
+          // When a click event occurs on a feature in the places layer, open a popup at the
+          // location of the feature, with description HTML from its properties.
+          map.on('click', 'places', function(e) {
+            var coordinates = e.features[0].geometry.coordinates.slice();
+
+            // Ensure that if the map is zoomed out such that multiple
+            // copies of the feature are visible, the popup appears
+            // over the copy being pointed to.
+            while (Math.abs(e.lngLat.lng - coordinates[0]) > 180) {
+              coordinates[0] += e.lngLat.lng > coordinates[0] ? 360 : -360;
+            }
+            var activeItem = document.querySelector('li.active');
+            if (activeItem) {
+              activeItem.classList.remove('active');
```

Comment:
Is this modifying something now managed by React?

Reply:
Yes, this is part of the React list that I create lower down. Should I not do this?

Reply:
Generally, when React owns part of the DOM, it should fully control it.  It internally maintains a shadow DOM and takes care of updating the page's DOM when things change.  Changes to any React DOM should happen via props/state changes managed by React.

Reply:
I think my new version fixes this